src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/FeatureCheckDataFlow.cs

```diff
@@ -0,0 +1,999 @@
+					RequiresDynamicCode ();
+			}
+
+			static void GuardedElse ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+				{
+					throw new Exception ();
+				}
+				else
+				{
+					RequiresDynamicCode ();
+				}
+			}
+
+			static void GuardedAnd ()
+			{
+				var a = RuntimeFeature.IsDynamicCodeSupported && RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedOr ()
+			{
+				var a = !RuntimeFeature.IsDynamicCodeSupported || RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedTernary ()
+			{
+				var a = RuntimeFeature.IsDynamicCodeSupported ? RequiresDynamicCodeBool () : true;
+				var b = !RuntimeFeature.IsDynamicCodeSupported ? true : RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedThrow ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+				{
+					throw new Exception ();
+				}
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedAssert ()
+			{
+				Debug.Assert (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfTrue ()
+			{
+				DoesNotReturnIfTrue (!RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfFalse ()
+			{
+				DoesNotReturnIfFalse (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturn ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+					DoesNotReturn ();
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfFalseCtor ()
+			{
+				new DoesNotReturnIfFalseCtor (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+		}
+
+		class FeatureCheckBooleanExpressions
+		{
+			[ExpectedWarning ("IL2026", nameof (RequiresUnreferencedCode), ProducedBy = Tool.Trimmer | Tool.NativeAot)]
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void And ()
+			{
+				if (RuntimeFeature.IsDynamicCodeSupported && TestFeatures.IsUnreferencedCodeSupported) {
+					RequiresDynamicCode ();
+					RequiresUnreferencedCode ();
+				}
+			}
+
```

Comment:
To get consistent behavior for trimmer we should pass the right feature switch so that IsDynamicCodeSupported is substituted to false. The NativeAOT already does this in the test infra for all tests: https://github.com/dotnet/runtime/blob/df6fdefa27068126794b253d4d822706221a92db/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/TrimmingArgumentBuilder.cs#L45C69-L45C69
This test case should also do it for the trimmer.

Similarly to emulate the equivalent behavior in AOT we should add a substitutions XML for the AOT tests which will set the TestFeature properties to constants (since AOT can't do constant prop from return values).

We should note the behavior differences, but for these tests we should try to make them behave as close to each other as possible since the goal of the test is to validate all the patterns which are correctly recognized by each tool - specifying feature switches and substitutions is not part of that I think.